src/clp_logging/handlers.py

```diff
@@ -792,3 +799,168 @@ def __init__(
         super().__init__(
             open(fpath, mode), enable_compression, timestamp_format, timezone, loglevel_timeout
         )
+
+
+class ClpKeyValuePairStreamHandler(logging.Handler):
+    """
+    A custom logging handler that serializes key-value pair log events into the
+    CLP key-value pair IR format.
+
+    Differences from `logging.StreamHandler`:
+      - Expects log events (`logging.LogRecord`) to include key-value pairs represented as a Python
+        dictionary.
+      - Serializes the key-value pairs into the CLP key-value pair IR format before writing to the
+        stream.
+
+    Rules for key-value pair representation:
+      - Key:
+        - Must be of type `str`.
+      - Value:
+        - Must be one of the following types:
+          - Primitive types: `int`, `float`, `str`, `bool`, or `None`.
+          - Arrays:
+            - May contain primitive values, dictionaries, or nested arrays.
+            - Can be empty.
+          - Dictionaries:
+            - Must adhere to the same key-value rules.
+            - Can be empty.
+
+    :param stream: A writable byte output stream to which the handler will write the serialized IR
+        byte sequences.
+    :param enable_compression: Whether to compress the serialized IR byte sequences using zstd.
+    """
+
+    def __init__(
+        self,
+        stream: IO[bytes],
+        enable_compression: bool = True,
+    ) -> None:
+        super().__init__()
+
+        self._enable_compression: bool = enable_compression
+        self._serializer: Optional[Serializer] = None
+        self._formatter: Optional[logging.Formatter] = None
+        self._ostream: IO[bytes] = stream
+
+        self._auto_gen_kv_pairs_buf: AutoGeneratedKeyValuePairsBuffer = (
+            AutoGeneratedKeyValuePairsBuffer()
+        )
+
+        self._init_new_serializer(stream)
+
+    # override
+    def setFormatter(self, fmt: Optional[logging.Formatter]) -> None:
+        if fmt is None:
+            return
+        warnings.warn(
+            f"Formatter is currently not supported in the current {self.__class__.__name__}",
+            category=RuntimeWarning,
+        )
+        self._formatter = fmt
+
+    # override
+    def emit(self, record: logging.LogRecord) -> None:
+        """
+        Overrides `logging.Handler.emit` to ensure `logging.Handler.handleError`
+        is always called and avoid requiring a `logging.LogRecord` to call
+        internal writing functions.
```

Comment:
```suggestion
        is always called.
```

Talked to David and he mentioned the "avoids requiring..." part probably doesn't make sense in newer Python versions. I think we can also refactor the docstring in `CLPBaseHandler.emit` to:

```python
        """
        Overrides `logging.Handler.emit` in base class to ensure
        `logging.Handler.handleError` is always called. Derived classes should only need to implement
        `_write` instead of overriding this method.
        """
```

Reply:
Looking at Python 3.10's implementation of `logging.StreamHandler`:
```Python
    def emit(self, record):
        """
        Emit a record.

        If a formatter is specified, it is used to format the record.
        The record is then written to the stream with a trailing newline.  If
        exception information is present, it is formatted using
        traceback.print_exception and appended to the stream.  If the stream
        has an 'encoding' attribute, it is used to determine how to do the
        output to the stream.
        """
        try:
            msg = self.format(record)
            stream = self.stream
            # issue 35046: merged two stream.writes into one.
            stream.write(msg + self.terminator)
            self.flush()
        except RecursionError:  # See issue 36272
            raise
        except Exception:
            self.handleError(record)
```
The default implementation will use `format` method to convert the record into a formatted string and directly write the string to the underlying stream. This also matches the "`logging.Handler.handleError` is always called" part. So, I guess this is what the original docstring was talking about.

However, we are directly deriving from `logging.Handler` who has its `emit` method abstract at least since Python 3.6: https://github.com/python/cpython/blob/3.6/Lib/logging/__init__.py#L842. I don't think our current docstring is relevant since we will need to implement this method anyways. So how about sth like:
```
        """
        Overrides `logging.Handler.emit` to ensure the given record is encoded using CLP IR format
        before writing to the underlying stream.
        """
```

Reply:
* For the `ClpKeyValuePairStreamHandler`, I think we can drop the docstring since as you said, we need to implement it and the base class' method covers the description.
* For `CLPBaseHandler`, my suggestion was wrong. What David meant and what I wanted to convey is:

  ```python
          """
          Implements `logging.Handler.emit` to ensure `logging.Handler.handleError`
          is always called and so derived classes only need to implement `_write`
          instead of implementing this method.
          """
  ```
    * Lint as necessary.

Reply:
Shall we defer the `CLPBaseHandler` to another PR? I can create an issue to track it.

Reply:
I think it's small enough to include in this PR, but I'm fine either way.

Reply:
Changed in this PR.